### DISCUSSION TITLE
Dealing with SCM operations

### DISCUSSION THREAD

**Discussion by peacememories** [ethersync/ethersync#83] [General]
At my current workplace we use VSCode Live Share a lot to facilitate remote (and sometimes even local) pair programming.

While developing we do two things fairly frequently:
- Commit (we tend to keep our changes very small so a work session can have dozens of commits)
- Use [Jujutsu](https://github.com/martinvonz/jj) to jump around commits and rewrite the history

Live Share, by its nature, handles this fairly well because no files on the client have to be changed.

For ethersync I see a couple of potential issues with this workflow.

When committing on one machine you could either:
- Not commit on the other machine. After the work session the peer would have to reset everything before continuing to work, even if they wanted to continue where they left off with their partner.
- Also perform commits on the other peers. This seems extremely complicated and would obviously only work with supported SCMs.
- Also copy the `.jj`, `.git`, `.svn`, or whatever directories, using that to keep SCM in sync. I don't know how brittle this is. It would definitely result in lots of changes being synced. This could also result in corruption especially with Jujutsu, since local "change IDs" do not match between different users.

When rewriting and jumping around history, lots of the same issues apply. Additionally this can result in massive amounts of changes. We regularly use a custom command `jj_run` which goes through _every commit_ in a chunk of history and runs a command (e.g. autoformatter).

I wonder if this workflow can be reconciled with how ethersync works. I would definitely love if we could manage that :sweat_smile:

**Comment by blinry** [comment]
When we "dogfeed" Ethersync ourselves, we are using the following workaround:

1. One of the people in the session makes a commit, and pushes it.
2. The other people run `git fetch`, and then ✨ `git reset --mixed @{u}` ✨ Which sets your current branch on the upstream commit, and also updates the index, without touching the files.

After that, both people are on the same commit, and the status for both of them is clean, which feels a bit like magic. :D

A better integration would be nice for sure, especially to allow people to change branches together. That's not our top priority right now, but I'd also love it to find good solutions there!

We decided we wanna try GitHub's "Discussions" feature, I'm moving this thread there for now! :)
